lib/puppet/configurer.rb

```diff
@@ -225,12 +225,26 @@ def run(options = {})
         # mode. We shouldn't try to do any failover in that case.
         if options[:catalog].nil? && do_failover
           server, port = find_functional_server
-          if server.nil?
-            raise Puppet::Error, _("Could not select a functional puppet master from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) }
-          else
-            #TRANSLATORS 'server_list' is the name of a setting and should not be translated
-            Puppet.debug _("Selected puppet server from the `server_list` setting: %{server}:%{port}") % { server: server, port: port }
-            report.master_used = "#{server}:#{port}"
+          begin
+            if server.nil?
+              raise Puppet::Error, _("Could not select a functional puppet server from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) }
+            else
+              #TRANSLATORS 'server_list' is the name of a setting and should not be translated
+              Puppet.debug _("Selected puppet server from the `server_list` setting: %{server}:%{port}") % { server: server, port: port }
+              report.master_used = "#{server}:#{port}"
+            end
+          rescue Puppet::Error => detail
```

Comment:
Sorry I missed this earlier, because I thought `find_functional_server` was within the `begin/rescue` clause. I'm not a huge fan of using exceptions for flow control. Could we just handle the `server.nil?` case directly instead of raising an exception only to rescue it again?

```ruby
server, port = find_functional_server
if server.nil?
  if Puppet[:usecacheonfailure]
    options[:pluginsync] = false
    @running_failure = true
    server = Puppet[:server_list].first[0]
    port = Puppet[:server_list].first[1] || Puppet[:masterport]
  else
    detail = Puppet::Error, _("Could not select a functional puppet server from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) }
    Puppet.log_exception(detail)
    raise detail
  end
end

Puppet.override(server: server, serverport: port) do
  ...
```

Reply:
Yes, removing exception code is simpler:
```
@@ -226,7 +226,18 @@ class Puppet::Configurer
         if options[:catalog].nil? && do_failover
           server, port = find_functional_server
           if server.nil?
-            raise Puppet::Error, _("Could not select a functional puppet master from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) }
+            detail = _("Could not select a functional puppet server from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) }
+            if Puppet[:usecacheonfailure]
+              options[:pluginsync] = false
+              @running_failure = true
+              if server.nil?
+                server = Puppet[:server_list].first[0]
+                port = Puppet[:server_list].first[1] || Puppet[:masterport]
+              end
+              Puppet.err(detail)
+            else
+              raise Puppet::Error, detail
+            end
           else
             #TRANSLATORS 'server_list' is the name of a setting and should not be translated
             Puppet.debug _("Selected puppet server from the `server_list` setting: %{server}:%{port}") % { server: server, port: port }

```